### PR TITLE
addAccountContext w currency and decimals

### PR DESCRIPTION
**Pull request by mejango**
# Description

Adding accounting contexts now allows clients to specify how many decimals the token does its accounting with and what currency it corresponds to. This puts more responsibility onto clients to get these values right, while allowing the protocol to adapt to edge cases better. These edge cases are 1) erc20 doesnt expose `decimals` function, and 2) two tokens share a first 4 bytes therefor have clashing currencies. 

## Limitations & risks

*Are there any trade-off or new vulnarbility surface based on theses changes?*

# Check-list
- [ ] Tests are covering the new feature
- [ ] Code is [natspec'd](https://docs.soliditylang.org/en/v0.8.17/natspec-format.html)
- [ ] Code is [linted and formatted](https://docs.soliditylang.org/en/v0.8.17/style-guide.html)
- [ ] I have run the test locally (and they pass)
- [ ] I have rebased to the latest main commit (and tests still pass)

# Interactions
These changes will impact the following contracts:
- Directly:

- Indirectly:

### CODE REVIEW

File: src/JBMultiTerminal.sol [hunk 1]
```diff
@@ -555,31 +576,43 @@ contract JBMultiTerminal is JBPermissioned, ERC2771Context, IJBMultiTerminal {
...
+
+            // Make sure the decimals provided in the accounting context matches expectations if there are expectations.
+            if (
+                (accountingContext.token == JBConstants.NATIVE_TOKEN && accountingContext.decimals != 18)
+                    || (
+                        IERC165(accountingContext.token).supportsInterface(type(IERC20Metadata).interfaceId)
+                            && accountingContext.decimals != IERC20Metadata(accountingContext.token).decimals()
+                    )
+            ) {
+                revert INVALID_ACCOUNTING_CONTEXT_DECIMALS();
```
Comment (simplemachine92) [inline src/JBMultiTerminal.sol:604]: LGTM except this. decimals can be manipulated by a custom impl (overridden) decimals() of an erc20. 

It's also kind of a gotcha from UX perspective that they could enter the wrong decimals when we could just grab that for them. 

So I say check for interface support -> grab decimals = reduce calldata and gotchas.

Comment (mejango) [reply src/JBMultiTerminal.sol:604]: ... we first check to see if the token reports `decimals()`, if so we make sure the provided value aligns.

Comment (mejango) [reply src/JBMultiTerminal.sol:604]: image

Comment (mejango) [reply src/JBMultiTerminal.sol:604]: the thing now is that we also allow the client to provide custom value in case the token doesn't report

**Pull request merged**